File: java/SudokuSolver.java
/*
 * Filename: NextPermutation.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Fri 20 Mar 2015 05:59:58 PM EDT
 */

public class Solution {
	public void solveSudoku(char[][] board) {
		if (board == null || board.length != 9 || board[0].length != 9) {
			return;
		}

		solveSudokuAux(board, 0, 0);
		return;
	}

	public static void solveSudokuAux(char[][] board, int i, int j) {
		if (j == 9) {
			j = 0;
			i++;
		}

		if (i == 9) {
			return;
		}

		int n = board.length;
		char[] token = new char[]{ '1', '2', '3', '4', '5', '6', '7', '8', '9' };

		if (board[i][j] == '.') {
			int iBlock = i / 3;
			int jBlock = j / 3;

			for (int k = 3 * iBlock; k < 3 * (iBlock + 1); k++) {
				for (int l = 3 * jBlock; l < 3 * (jBlock + 1); l++) {
					if (board[k][l] >= '1' && board[k][l] <= '9') {
						token[board[k][l] - '1'] = '0';
					}
				}
			}

			for (int k = 0; k < n; k++) {
				if (k != i && board[k][j] >= '1' && board[k][j] <= '9') {
					token[board[k][j] - '1'] = '0';
				}
				if (k != j && board[i][k] >= '1' && board[i][k] <= '9') {
					token[board[i][k] - '1'] = '0';
				}
			}

			for (int k = 0; k < token.length; k++) {
				if (token[k] != '0') {
					board[i][j] = token[k];
					solveSudokuAux(board, i, j+1);
					if (isFull(board)) {
						return;
					}
					board[i][j] = '.';
				}
			}
		} else {
			solveSudokuAux(board, i, j+1);
		}

		return;
	}

	private boolean isFull(char[][] board) {
		for (int i = 0; i < board.length; i++) {
			for (int j = 0; j < board[i].length; j++) {
				if (board[i][j] > '9' || board[i][j] < '1') {
					return false;
				}
			}
		}
		return true;
	}
}

File: cpp/sudoku_solver.cpp
/*
 * Filename: sudoku_solver.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Tue 03 Nov 2015 09:23:11 PM EST
 * Description: 
 */

#include <iostream>

using namespace std;

class Solution {
	public:
		void solveSudoku(vector<vector<char> >& board)
		{
			if (board.size() != 9 || board[0].size() != 9)
			{
				return;
			}

			solvableFrom(board, 0, 0);
		}

private:
		bool solvableFrom(
				vector<vector<char> >& board,
				int r,
				int c)
		{
			if (r == 9)
			{
				return true;
			}
			else if (c == 9)
			{
				return solvableFrom(board, r+1, 0);
			}
			else if (board[r][c] != '.')
			{
				return solvableFrom(board, r, c+1);
			}
			else
			{
				for (char p = '1'; p <= '9'; p++)
				{
					if (!isValid(board, r, c, p))
					{
						continue;
					}

					board[r][c] = p;
					if (solvableFrom(board, r, c+1) &&
							isFull(board))
					{
						return true;
					}
				}

				board[r][c] = '.';
			}
		}

		bool isValid(
				vector<vector<char> >& board,
				int r,
				int c,
				char p)
		{
			for (int i = 0; i < 9; i++)
			{
				if (board[i][c] == p || board[r][i] == p)
				{
					return false;
				}
			}

			int index = r / 3 * 3 + c / 3;
			for (int i = 0; i < 3; i++)
			{
				for (int j = 0; j < 3; j++)
				{
					if (board[index / 3 * 3 + i][index % 3 * 3 + j] == p)
					{
						return false;
					}
				}
			}

			return true;
		}

		bool isFull(
				vector<vector<char> >& board)
		{
			for (auto i = 0; i < 9; i++)
			{
				for (auto j = 0; j < 9; j++)
				{
					if (board[i][j] < '0' || board[i][j] > '9')
					{
						return false;
					}
				}
			}

			return true;
		}
};
